From b4be3e4660f64bc6cdfd6505240741bfbede936b Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Tue, 07 Jun 2022 10:08:12 +0300
Subject: [PATCH] [SW-89972] gaudi3: erom: load all images from erom image descriptors

Currently, ROM as part of EROM images bringup reads and load PCIe
and RECOVERY image from image descriptor. This patch adds flexible
approach and loads all the images as part of EROM image descriptor.

Change-Id: Ic532b24a4b75d04b60de59bae4228c933801fe93
---

diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index 532ee7e..5b59b66 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -25,9 +25,9 @@
 	return DMA_DEST_HBW;
 }
 
-void ext_rom_load_erom_image(enum hl_fw_img_type_erom img_type)
+void ext_rom_load_erom_image(uint32_t index)
 {
-	enum arcboot_rom_progress img_sts_base = PROGRESS_ARCBOOT_EROM_PCIE;
+	enum arcboot_rom_progress img_sts_base = PROGRESS_ARCBOOT_EROM_IMG0;
 	enum dma_error dma_err = DMA_SUCCESS;
 	uint32_t idx, num_crc, img_len_nocrc;
 	enum dma_dest_type dst_type;
@@ -37,18 +37,17 @@
 	void (*erom_func)();
 	int err;
 
-	if (img_type == HL_FW_EROM_RECOVERY)
-		img_sts_base = PROGRESS_ARCBOOT_EROM_RECOVERY;
+	img_sts_base = img_sts_base + (index *
+			((PROGRESS_ARCBOOT_EROM_IMG0_FINISH - PROGRESS_ARCBOOT_EROM_IMG0) + 1));
 
 	arcboot_rom_status_update_bringup(img_sts_base, PROGRESS_EROM_BRINGUP_START);
 
-	idx = boot_desc_get_img_desc(HL_FW_BOOT_PTR_EROM, img_type, &desc, FLOPS_UPD_NOOP);
+	idx = boot_desc_get_img_desc_by_index(HL_FW_BOOT_PTR_EROM, index, &desc, FLOPS_UPD_NOOP);
 	if (idx == HL_FW_IMG_NOT_FOUND)
 		return;
 
 	dst_type = arcboot_rom_get_dma_dest_type(desc.loadaddr);
 
-	/* TODO: check image descriptor values */
 	arcboot_rom_status_update_bringup(img_sts_base, PROGRESS_EROM_DMA_START);
 
 	dma_err = dma_copy_image_v2(desc.img_size, desc.spi_offset,
@@ -98,23 +97,22 @@
 
 static void execute_erom_recovery(void)
 {
+	uint32_t count, i;
+
 	/* EROM recovery not supported in slave die */
 	if (arcboot_rom_global.die_id)
 		return;
 
 	arcboot_rom_status_update(PROGRESS_ARCBOOT_IN_EROM_RECOVERY);
 
-	if (!boot_desc_init(HL_FW_BOOT_PTR_EROM)) {
-		arcboot_rom_status_update(
-				PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR);
+	count = boot_desc_init(HL_FW_BOOT_PTR_EROM);
+	if (!count) {
+		arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR);
 		return;
 	}
 
-	/* Load PCIE binary from eROM to establish the link */
-	ext_rom_load_erom_image(HL_FW_EROM_PCIE);
-
-	/* Load RECOVERY binary from eROM */
-	ext_rom_load_erom_image(HL_FW_EROM_RECOVERY);
+	for (i = 0; i < count; i++)
+		ext_rom_load_erom_image(i);
 }
 
 static void rom_pcie_recovery(void)
diff --git a/include/arcboot.h b/include/arcboot.h
index d5ca802..a099b94 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -198,50 +198,60 @@
 	PROGRESS_ARCBOOT_RAM_START = 65,
 	PROGRESS_ARCBOOT_RECOVERY = 66,
 	PROGRESS_ARCBOOT_RECOVERY_DMA = 67,
-	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 68,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_START = 69,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_UNALIGNED_DEST = 70,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 71,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_ERROR = 72,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_FINISH = 73,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 74,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_VERIFICATION_FAIL = 75,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR = 76,
-	PROGRESS_ARCBOOT_EROM_PCIE = 77,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_START = 78,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_CRC_FAILED = 79,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_UNALIGNED_DEST = 80,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_UNALIGNED_IMG_SIZE = 81,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_ERROR = 82,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_FINISH = 83,
-	PROGRESS_ARCBOOT_EROM_PCIE_VERIFICATION_FAIL = 84,
-	PROGRESS_ARCBOOT_EROM_PCIE_START = 85,
-	PROGRESS_ARCBOOT_EROM_PCIE_FINISH = 86,
-	PROGRESS_ARCBOOT_EROM_RECOVERY = 87,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_START = 88,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_CRC_FAILED = 89,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_UNALIGNED_DEST = 90,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_UNALIGNED_IMG_SIZE = 91,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_ERROR = 92,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_FINISH = 93,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_VERIFICATION_FAIL = 94,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_START = 95,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH = 96,
-	PROGRESS_ARCBOOT_EROM_TPM = 97,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_START = 98,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_CRC_FAILED = 99,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_UNALIGNED_DEST = 100,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_UNALIGNED_IMG_SIZE = 101,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_ERROR = 102,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH = 103,
-	PROGRESS_ARCBOOT_EROM_TPM_VERIFICATION_FAIL = 104,
-	PROGRESS_ARCBOOT_EROM_TPM_START = 105,
-	PROGRESS_ARCBOOT_EROM_TPM_FINISH = 106,
-	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 107,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 108,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 109,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 110,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 111,
+	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 68,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 69,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 70,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 71,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 72,
+	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 73,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_START = 74,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_UNALIGNED_DEST = 75,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 76,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_ERROR = 77,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_FINISH = 78,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 79,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_VERIFICATION_FAIL = 80,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR = 81,
+	PROGRESS_ARCBOOT_EROM_PCIE = 82,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_START = 83,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_CRC_FAILED = 84,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_UNALIGNED_DEST = 85,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_UNALIGNED_IMG_SIZE = 86,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_ERROR = 87,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_FINISH = 88,
+	PROGRESS_ARCBOOT_EROM_PCIE_VERIFICATION_FAIL = 89,
+	PROGRESS_ARCBOOT_EROM_PCIE_START = 90,
+	PROGRESS_ARCBOOT_EROM_PCIE_FINISH = 91,
+	PROGRESS_ARCBOOT_EROM_RECOVERY = 92,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_START = 93,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_CRC_FAILED = 94,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_UNALIGNED_DEST = 95,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_UNALIGNED_IMG_SIZE = 96,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_ERROR = 97,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_FINISH = 98,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_VERIFICATION_FAIL = 99,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_START = 100,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH = 101,
+	PROGRESS_ARCBOOT_EROM_TPM = 102,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_START = 103,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_CRC_FAILED = 104,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_UNALIGNED_DEST = 105,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_UNALIGNED_IMG_SIZE = 106,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_ERROR = 107,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH = 108,
+	PROGRESS_ARCBOOT_EROM_TPM_VERIFICATION_FAIL = 109,
+	PROGRESS_ARCBOOT_EROM_TPM_START = 110,
+	PROGRESS_ARCBOOT_EROM_TPM_FINISH = 111,
+	PROGRESS_ARCBOOT_EROM_IMG0 = 112,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 113,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 114,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 115,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 116,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 117,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 118,
+	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 119,
+	PROGRESS_ARCBOOT_EROM_IMG0_START = 120,
+	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 121,
 };
 
 enum security_location {
